internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -320,159 +261,67 @@ func (bt *bucketTicker) add(bucketID string) {
-		currCap:       initCapacity,
-		nextEmptyItem: uint32(0),
-		sendItemsSize: sendBatchSize,
-
-		closed: uint32(0),
-	}
-
-	batch.items.Store(make([]*tracepb.Span, initCapacity))
-	return batch
+	bt.once.Do(bt.runTicker)
 }
 
-func (b *batch) add(spans []*tracepb.Span) (cut bool, closed bool) {
-	if atomic.LoadUint32(&b.closed) == batchStatusClosed {
-		return false, true
-	}
-	// Add to pending and then check closed again. If closed, just return and decrement pending.
-	atomic.AddInt32(&b.pending, int32(1))
-	defer atomic.AddInt32(&b.pending, int32(-1))
-	if atomic.LoadUint32(&b.closed) == batchStatusClosed {
-		return false, true
-	}
-	openTill := atomic.AddUint32(&b.nextEmptyItem, uint32(len(spans)))
-	openFrom := openTill - uint32(len(spans))
-	currCap := atomic.LoadUint32(&b.currCap)
-	if openTill > currCap {
-		b.grow(openTill)
-	}
-
-	for spanIndex, span := range spans {
-		b.items.Load().([]*tracepb.Span)[openFrom+uint32(spanIndex)] = span
-	}
-
-	return openTill > b.sendItemsSize, false
-}
-
-func (b *batch) closeBatch() {
-	atomic.StoreUint32(&b.closed, batchStatusClosed)
+func (bt *bucketTicker) runTicker() {
 	for {
-		// We only need to wait for goroutines currently executing `add`
-		// We are safe from closing during a items mutation due to the second
-		// closed check after incrementing pending
-		pending := atomic.LoadInt32(&b.pending)
-		if pending == int32(0) {
+		select {
+		case <-bt.ticker.C:
+			for nbKey := range bt.nodes {
+				nb := bt.parent.getBucket(nbKey)
+				// Need to check nil here incase the node was deleted from the parent batcher, but
+				// not deleted from bt.nodes
+				if nb == nil {
+					// re-delete just in case
+					delete(bt.nodes, nbKey)
+					continue
+				} else {
+					bt.processNodeBatch(nbKey, nb)
+				}
+			}
+		case newBucketKey := <-bt.pendingNodes:
+			bt.nodes[newBucketKey] = true
+		case <-bt.stopCn:
+			bt.ticker.Stop()
 			return
 		}
-		runtime.Gosched()
 	}
 }
 
-func (b *batch) grow(neededSize uint32) {
-	b.growMu.Lock()
-	defer b.growMu.Unlock()
-	currCap := atomic.LoadUint32(&b.currCap)
-	// If we entered this function concurrently with another call to grow,
-	// make sure we don't needlessly re-copy items
-	if neededSize < currCap {
-		return
-	}
-	newCap := currCap
-	for newCap < neededSize {
-		newCap = newCap * 2
-	}
-	newItems := make([](*tracepb.Span), newCap, newCap)
-	copy(newItems, b.items.Load().([]*tracepb.Span))
-	b.items.Store(newItems)
-	// It is important that we store the new cap after copying all items, so
-	// that concurrent adds also attempt to grow and acquire the lock
-	atomic.StoreUint32(&b.currCap, newCap)
-}
+func (bt *bucketTicker) processNodeBatch(nbKey string, nb *nodeBatch) {
+	nb.mu.Lock()
+	if nb.totalItemCount > 0 {
+		// If the batch is non-empty, go ahead and send it
+		itemCount := nb.totalItemCount
+		var itemsToProcess [][]*tracepb.Span
+		if nb.lastSent+bt.parent.timeout.Nanoseconds() < time.Now().UnixNano() {
+			itemsToProcess = nb.items
```

Comment:
This seems like a good place to avoid duplicate code between this function and `func (nb *nodeBatch) add`

```go
func (nb *nodeBatch) getAndReset() ([][]*tracepb.Span, int){
  itemsToProcess := nb.items
  itemsCount := nb.totalItemCount
  nb.items = make([][]*tracepb.Span, 0, initialBatchCapacity)
  nb.lastSent = time.Now().UnixNano()
  nb.totalItemCount = 0
  return itemsToProcess, itemsCount
}
```